### PR TITLE
feat: dynamic schema update

### PR DESCRIPTION
**Pull request by sdorokhova**
## Description

- on Operate startup, after index versions are validated and data migration finished (old style), index fields will be validated and new fields added
- add position field to required indices
- add additional not-indexed `position` fields (explained [here](https://github.com/camunda/operate/issues/6477#issuecomment-2017533049))
- add some default values for older field definitions in order for field comparison to succeed
- unit tests + extended migration test suite

## Related issues

Related with https://github.com/camunda/operate/issues/6483

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: operate/common/src/main/java/io/camunda/operate/conditions/DatabaseInfo.java
Comment (ThorbenLindhauer) [inline operate/common/src/main/java/io/camunda/operate/conditions/DatabaseInfo.java:68]: Maybe it's a bit nicer to put this logic into `OperateProperties`?

File: operate/common/src/main/java/io/camunda/operate/util/JSONUtil.java
Comment (ThorbenLindhauer) [inline operate/common/src/main/java/io/camunda/operate/util/JSONUtil.java:26]: This class is specific to Elasticsearch, so maybe we can include that in the name

File: operate/common/src/main/java/io/camunda/operate/util/JSONUtil.java
Comment (ThorbenLindhauer) [inline operate/common/src/main/java/io/camunda/operate/util/JSONUtil.java:28]: Since this reads only classpath resources and not any file, could we name the method accordingly?

File: operate/importer-8_5/src/main/java/io/camunda/operate/zeebeimport/v8_5/processors/ListViewZeebeRecordProcessor.java
Comment (ThorbenLindhauer) [inline operate/importer-8_5/src/main/java/io/camunda/operate/zeebeimport/v8_5/processors/ListViewZeebeRecordProcessor.java:249]: As we would merge this PR individually of the follow-up ticket, I suggest to leave the TODO out of the code and make a note in the follow-up ticket.

File: operate/schema/src/main/java/io/camunda/operate/schema/IndexMappingDifference.java
Comment (ThorbenLindhauer) [inline operate/schema/src/main/java/io/camunda/operate/schema/IndexMappingDifference.java:153]: Is it necessary that this returns a function or could this method just implement the plain logic and you use a function reference to it?

File: operate/schema/src/main/java/io/camunda/operate/schema/IndexSchemaValidator.java
Comment (ThorbenLindhauer) [inline operate/schema/src/main/java/io/camunda/operate/schema/IndexSchemaValidator.java:103]: What about the FIXME comment? Do you plan to do that as part of this PR? Should there be a ticket for this?

File: operate/schema/src/main/java/io/camunda/operate/schema/IndexSchemaValidator.java
Comment (ThorbenLindhauer) [inline operate/schema/src/main/java/io/camunda/operate/schema/IndexSchemaValidator.java:158]: You could extract methods for the big building blocks of this method to make it better readable

File: operate/schema/src/main/java/io/camunda/operate/schema/IndexSchemaValidator.java
Comment (ThorbenLindhauer) [inline operate/schema/src/main/java/io/camunda/operate/schema/IndexSchemaValidator.java:205]: Should this be logged as warning if we consider it expected behavior in rolling update scenarios? Warning will let people believe that something is wrong, and the message doesn't help them resolve it. So I'd suggest to go either way: Choose another log level or give more information what this message means and what to do with it.

File: operate/schema/src/main/java/io/camunda/operate/schema/IndexSchemaValidator.java
Comment (ThorbenLindhauer) [inline operate/schema/src/main/java/io/camunda/operate/schema/IndexSchemaValidator.java:236]: Should we move this to `IndexDescriptor` or its subclasses so that all the naming patterns are in one place?

File: operate/schema/src/main/java/io/camunda/operate/schema/opensearch/OpensearchSchemaManager.java
Comment (ThorbenLindhauer) [inline operate/schema/src/main/java/io/camunda/operate/schema/opensearch/OpensearchSchemaManager.java:257]: There are constants in this class that you could use for the Strings. Also, you can consider putting this into the `IndexDescriptor` and `TemplateDescriptor` classes for a more object-oriented approach and less if/else.

File: operate/schema/src/main/java/io/camunda/operate/schema/opensearch/OpensearchSchemaManager.java
Comment (ThorbenLindhauer) [inline operate/schema/src/main/java/io/camunda/operate/schema/opensearch/OpensearchSchemaManager.java:263]: I suggest to either use the files' actual charset or just pass the `InputStream` to Jackson. Same pattern is in `ElasticsearchSchemaManager`.

File: operate/schema/src/main/java/io/camunda/operate/store/elasticsearch/RetryElasticsearchClient.java
Comment (ThorbenLindhauer) [inline operate/schema/src/main/java/io/camunda/operate/store/elasticsearch/RetryElasticsearchClient.java:218]: What's the purpose of this comment?

File: operate/schema/src/main/java/io/camunda/operate/store/opensearch/client/sync/OpenSearchIndexOperations.java
Comment (ThorbenLindhauer) [inline operate/schema/src/main/java/io/camunda/operate/store/opensearch/client/sync/OpenSearchIndexOperations.java:222]: This method is not used and the other one should be renamed.

File: operate/schema/src/main/java/io/camunda/operate/schema/IndexSchemaValidator.java
Comment (sdorokhova) [reply operate/schema/src/main/java/io/camunda/operate/schema/IndexSchemaValidator.java:103]: Since we still left the possibility to do old style migration, this comment is not valid. I will remove it.

File: operate/schema/src/main/java/io/camunda/operate/store/elasticsearch/RetryElasticsearchClient.java
Comment (sdorokhova) [reply operate/schema/src/main/java/io/camunda/operate/store/elasticsearch/RetryElasticsearchClient.java]: Removed) Some leftovers

File: operate/schema/pom.xml
Comment (matt-whiteman) [inline operate/schema/pom.xml:123]: I think we should keep changes related to running our ITs under opensearch as a separate effort. There is other work that needs to be done as part of this, as well as the IT refactoring effort, and it would be good to keep changes matched with individual tasks and epics

File: operate/qa/migration-tests/migration-test/src/test/java/io/camunda/operate/qa/migration/util/AbstractMigrationTest.java
Comment (matt-whiteman) [inline operate/qa/migration-tests/migration-test/src/test/java/io/camunda/operate/qa/migration/util/AbstractMigrationTest.java:59]: These types of tests where we only load certain classes have been very hard to maintain in the past as new components and dependencies get added. Is there any alternative approach rather than specifying each individual class?

File: operate/qa/util/src/main/java/io/camunda/operate/qa/util/TestContainerUtil.java
Comment (matt-whiteman) [inline operate/qa/util/src/main/java/io/camunda/operate/qa/util/TestContainerUtil.java:96]: I recommend keeping the comment that was associated with this property. It has been changed several times by other people and breaks the test containers each time, the comment is meant to explain why.

File: operate/schema/src/test/java/io/camunda/operate/schema/IndexSchemaValidatorIT.java
Comment (matt-whiteman) [inline operate/schema/src/test/java/io/camunda/operate/schema/IndexSchemaValidatorIT.java:49]: I have concerns about making ITs like this that only load certain classes. In addition to the difficulty of maintaining (as we add new dependencies and components, these types of tests randomly break and need the class list expanded), it only causes a small part of the application to be loaded and can miss issues. Is there an issue with loading the application with the same full context as our other ITs?

**Comment by ThorbenLindhauer** [comment]
Thanks for the review @matt-whiteman, I will leave some comments to your questions about the tests since I wrote them to a large degree. That said, I will leave it to Svetlana and you to make the decisions since she is the DRI and both of you are the codebase owners.

File: operate/schema/src/test/java/io/camunda/operate/schema/IndexSchemaValidatorIT.java
Comment (ThorbenLindhauer) [reply operate/schema/src/test/java/io/camunda/operate/schema/IndexSchemaValidatorIT.java:63]: My idea was to have them in the schema module so that when you develop the module further, you can build and test the module on its own and already get some test coverage. Having them in the QA module would mean you have to build most of the operate modules and potentially wait for other unrelated tests to run too.

I don't have strong feelings in either direction, it is also "just" a matter of how you want to organize the tests.

File: operate/schema/src/test/java/io/camunda/operate/schema/IndexSchemaValidatorIT.java
Comment (ThorbenLindhauer) [reply operate/schema/src/test/java/io/camunda/operate/schema/IndexSchemaValidatorIT.java:49]: I think I started initially with the full application context approach and it became a bit frustrating, but I don't remember the details. I think problems were that there was no ready-made Spring Boot application (in the schema module) that creates the full context for the schema. And using package scanning, it was a situation that one thing only led to the next, with more and more beans unrelated to the test getting added in, requiring more configuration and doing more unrelated `@PostConstruct` logic.

That said, I also agree with your points. The current design was the in my opinion better tradeoff of having focused tests in the module at the expense of maintainability. Here as well, you could go in either direction.

File: operate/schema/pom.xml
Comment (ThorbenLindhauer) [reply operate/schema/pom.xml:123]: Fair point. Just to make sure we mean the same thing: I understand it that we still make this particular change now, just have separate commits/PRs and link it with a different ticket, right?

Even if it is only a small part of the Opensearch testing story, this profile helped us already quite a bit with developing as we found a bug in the Opensearch code that we would have otherwise found only later (e.g. after release like the one Ralf found recently).

Comment (matt-whiteman) [reply operate/schema/pom.xml:123]: That makes sense. If it's useful as part of the testing, that's fine. I was more concerned we were also pulling in the whole task for re-running all the tests with Opensearch as part of the pipeline and wanted to keep that separate since this is already a large effort. If this profile though is useful for testing and for this task, it makes perfect sense thank you.

File: operate/schema/src/test/java/io/camunda/operate/schema/IndexSchemaValidatorIT.java
Comment (matt-whiteman) [reply operate/schema/src/test/java/io/camunda/operate/schema/IndexSchemaValidatorIT.java:63]: I'm actually fine with keeping them in the schema module, I think it makes sense to have tests related to a specific module in that module rather than grouping them all under one subfolder. I just wanted to raise the question and see where thoughts were.

File: operate/schema/src/test/java/io/camunda/operate/schema/IndexSchemaValidatorIT.java
Comment (matt-whiteman) [reply operate/schema/src/test/java/io/camunda/operate/schema/IndexSchemaValidatorIT.java:49]: If it's too complex to run with a full application context, then this approach makes sense. I just wanted to raise the issue because especially for tests that are not run as part of the main pipeline (performance, migration, etc), if they break because of a new dependency added, it might not get caught for a while. We had situations where tests like these were found to be broken during a release just because of some changed dependencies.

**Comment by matt-whiteman** [comment]
> Thanks for the review @matt-whiteman, I will leave some comments to your questions about the tests since I wrote them to a large degree. That said, I will leave it to Svetlana and you to make the decisions since she is the DRI and both of you are the codebase owners.

Thanks @ThorbenLindhauer , left a couple replies. Your comments make sense, I just wanted to raise the issues to see if we considered alternatives.

File: operate/qa/migration-tests/migration-test/src/test/java/io/camunda/operate/qa/migration/util/AbstractMigrationTest.java
Comment (sdorokhova) [reply operate/qa/migration-tests/migration-test/src/test/java/io/camunda/operate/qa/migration/util/AbstractMigrationTest.java:59]: Yes, this is not ideal :( The thing is that we don't see much sense in doing unit tests for single classes and methods as the whole logic is in interaction with Elasticsearch and we would like to check behavior of our changes together with Elasticsearch, which brings more classes into play. We also don't want to use the whole Operate application context as then we will have our real Operate indices with much more complex structure and the tests won't be easy to read and maintain. With the current codebase I don't see how we could not list the classes to achieve the desired test behaviour. Do you think we should still switch on starting up the whole application context?

**Comment by sdorokhova** [comment]
Migration job run is green with latest changes: https://github.com/camunda/zeebe/actions/runs/8423262676/job/23064468550

File: operate/common/src/main/java/io/camunda/operate/connect/OpensearchConnector.java
Comment (matt-whiteman) [inline operate/common/src/main/java/io/camunda/operate/connect/OpensearchConnector.java:99]: Are we creating a bean of this type somewhere else as well? Primary can squash multiple bean issues but can lead to wrong expected bean types and usually means the creation needs to be looked at closer.

File: operate/common/src/main/java/io/camunda/operate/property/OperateProperties.java
Comment (matt-whiteman) [inline operate/common/src/main/java/io/camunda/operate/property/OperateProperties.java:379]: At some point in the future, we may want to create a wrapper component that holds the Elastic/Opensearch properties and makes the call to DatabaseInfo. This would allow checks for db type to be contained in one place and all client components just make an abstract property get call.

File: operate/common/src/main/java/io/camunda/operate/util/ElasticsearchJSONUtil.java
Comment (matt-whiteman) [inline operate/common/src/main/java/io/camunda/operate/util/ElasticsearchJSONUtil.java:28]: Suggestion for the future is to avoid static utility functions (or wrap them in a component that can be autowired), since static can't be mocked with Mockito.

File: operate/schema/src/main/java/io/camunda/operate/schema/IndexMappingDifference.java
Comment (matt-whiteman) [inline operate/schema/src/main/java/io/camunda/operate/schema/IndexMappingDifference.java:26]: Suggest some documentation on this class, as an outsider I'm not really sure under what circumstances it would be used or why

File: operate/schema/src/main/java/io/camunda/operate/schema/IndexSchemaValidator.java
Comment (matt-whiteman) [inline operate/schema/src/main/java/io/camunda/operate/schema/IndexSchemaValidator.java:47]: Suggest updating to constructor dependency injection wherever possible rather than autowiring member variables

File: operate/schema/src/test/java/io/camunda/operate/schema/util/opensearch/TestOpenSearchConnector.java
Comment (matt-whiteman) [inline operate/schema/src/test/java/io/camunda/operate/schema/util/opensearch/TestOpenSearchConnector.java:37]: Some of the logic in the test classes for opensearch and elasticsearch looks very similar and might want to consider an abstract class with common base logic they inherit from

**Review by matt-whiteman — APPROVED** [review]

**Comment by matt-whiteman** [comment]
CI failures need to be addressed before merging please

**Pull request merged**
